File: DictionaryProject/WordBuilder/ProjectSerializer.vb
Public Delegate Function LineParserDelegate(ByVal context As Object, ByVal reader As IO.TextReader, ByVal line As String, ByRef lineNumber As Integer) As Boolean

Public NotInheritable Class ProjectSerializer
    Private Sub New()
    End Sub

    Public Shared Function LoadDictionary(ByVal path As String) As DictionaryProject
        Dim result As New DictionaryProject()

        Using tr As New IO.StreamReader(path, Text.Encoding.UTF8)
            ReadDictionary(result, tr)
        End Using

        Return result
    End Function

    Private Shared Sub ReadDictionary(ByVal dic As DictionaryProject, ByVal reader As IO.TextReader)
        If Not ReadLines(dic, reader, New LineParserDelegate(AddressOf ParseDictionaryLine)) Then
            Throw New ApplicationException("Invalid dictionaryFile")
        End If
    End Sub

    Private Shared Function ParseDictionaryLine(ByVal context As Object, ByVal reader As IO.TextReader, ByVal line As String, ByRef lineNumber As Integer) As Boolean
        Dim parts As List(Of String) = ReadTokens(line)

        If parts.Count <> 2 OrElse parts(1) <> "{" Then
            Throw New ApplicationException(String.Format("Line {0}: Invalid dictionary file. Meaning lines must contain meaning followed by {{", lineNumber))
        End If

        Dim dic As DictionaryProject = DirectCast(context, DictionaryProject)

        Dim ctx As Context = ReadContext(reader, lineNumber)

        dic.AddWord(parts(0), ctx)

        Return True
    End Function

    Private Shared Function ReadContext(ByVal reader As IO.TextReader, ByVal lineNumber As Integer) As Context
        Dim result As New Context()

        If ReadLines(result, reader, AddressOf ParseContextLines, lineNumber) Then
            Throw New ApplicationException("End of file reached while reading a context.")
        End If

        Return result
    End Function

    Public Shared Sub SaveDictionary(ByVal dic As DictionaryProject, ByVal path As String)
        Using tw As New IO.StreamWriter(path, False, Text.Encoding.UTF8)
            For Each word In dic.Words.Keys
                For Each ctx As Context In dic.Words(word)
                    tw.WriteLine(String.Format("{0} {{", word))

                    WriteContext(ctx, tw)

                    tw.WriteLine("}")
                Next
            Next
        End Using
    End Sub

    Private Shared Sub WriteContext(ByVal context As Context, ByVal tw As IO.TextWriter, Optional ByVal indent As String = "")
        tw.WriteLine(String.Format("{0}{1}", indent, SecureList(context.Tokens)))

        For Each b As String In context.Branches.Keys
            tw.WriteLine(String.Format("{0}. {1} {{", indent, SecureString(b)))

            WriteContext(context.Branches(b), tw, indent & "  ")

            tw.WriteLine(String.Format("{0}}}", indent))
        Next

        For Each m As String In context.Marks.Keys
            tw.WriteLine(String.Format("{0}! {1} {2}", indent, SecureString(m), SecureString(context.Marks(m))))
        Next
    End Sub

    Private Shared Function ParseContextLines(ByVal context As Object, ByVal reader As IO.TextReader, ByVal line As String, ByRef lineNumber As Integer) As Boolean
        Dim parts As List(Of String) = ReadTokens(line)

        Dim ctx As Context = DirectCast(context, Context)

        If parts(0).Equals(".") Then
            If parts.Count <> 3 OrElse parts(2) <> "{" Then
                Throw New ApplicationException(String.Format("Line {0}: Invalid dictionary file. Branch lines must contain . followed by branch name followed by {{", lineNumber))
            End If
            Dim branch As Context = ctx.Branch(parts(1))
            branch.Tokens.Clear()
            If ReadLines(branch, reader, AddressOf ParseContextLines, lineNumber) Then
                Throw New ApplicationException("End of file reached while reading a context.")
            End If
        ElseIf parts(0).Equals("!") Then
            If parts.Count <> 3 Then
                Throw New ApplicationException(String.Format("Line {0}: Invalid dictionary file. Mark lines must contain ! followed by mark name followed by value.", lineNumber))
            End If
            ctx.Mark(parts(1), parts(2))
        ElseIf line.Equals("}") Then
            Return False
        Else
            ctx.Tokens.AddRange(parts)
        End If

        Return True
    End Function

    Public Shared Function LoadProject(ByVal path As String) As Project
        Dim result As New Project()

        Try
            Using tr As New IO.StreamReader(path, Text.Encoding.UTF8)
                ReadProject(result, tr)
            End Using
        Catch ex As Exception
            result.Warnings.Add(ex.Message)
        End Try

        Return result
    End Function

    Public Shared Function LoadProjectString(ByVal code As String) As Project
        Dim result As New Project

        Try
            Using tr As New IO.StringReader(code)
                ReadProject(result, tr)
            End Using
        Catch ex As Exception
            result.Warnings.Add(ex.Message)
        End Try

        Return result
    End Function

    Public Shared Function ReadLines(ByVal context As Object, ByVal reader As IO.TextReader, ByVal parser As LineParserDelegate, Optional ByRef lineNumber As Integer = 0) As Boolean
        Dim line As String = reader.ReadLine()
        While line IsNot Nothing
            line = line.Trim(" "c, Chr(9))
            lineNumber += 1
            If line.StartsWith("/*") Then
                Dim startingLine As Integer = lineNumber
                While line IsNot Nothing AndAlso Not line.EndsWith("*/")
                    line = reader.ReadLine()
                    lineNumber += 1
                End While
                If line Is Nothing Then
                    Throw New ApplicationException(String.Format("Line {0}: Block comment not closed correctly.", startingLine))
                End If
            ElseIf line.StartsWith("//") Then
                ' Ignore single line comments.
            ElseIf Not parser.Invoke(context, reader, line.Trim(" "c, Chr(9)), lineNumber) Then
                Return False
            End If
            line = reader.ReadLine()
        End While

        Return True
    End Function

    Public Shared Function ReadTokens(ByVal input As String) As List(Of String)
        Dim result As New List(Of String)()
        Dim start As Integer = 0
        Dim expand As New List(Of String)
        Dim didExpand As Boolean = False

        While start > -1
            Dim token As String = ReadToken(input, start, expand, didExpand)

            If didExpand Then
                result.AddRange(expand)
                expand.Clear()
            Else
                result.Add(token)
            End If
        End While

        Return result
    End Function

    Public Shared Function ExpandToken(ByVal input As String, ByVal origin As Integer, ByRef start As Integer, ByRef expandedToken As List(Of String), ByRef didExpand As Boolean, ByRef remove As Boolean) As String
        Dim token As String = input.Substring(origin, start - origin)

        If token.Contains("[") Then
            Dim nextOpen As Integer = input.IndexOf("["c, start)
            start = input.IndexOf("]"c, start - 1) + 1

            While nextOpen > -1 AndAlso nextOpen < start - 1
                nextOpen = input.IndexOf("["c, nextOpen + 1)
                start = input.IndexOf("]"c, start) + 1
            End While

            If start = 0 Then
                Throw New ApplicationException("Read token past end of line. Missing a ], are we?")
            End If

            token = input.Substring(origin, start - origin)
            Dim parts() As String = token.Substring(0, token.Length - 1).Split(New Char() {"["c}, 2)

            Dim copies As Integer = 0
            If expandedToken IsNot Nothing AndAlso Integer.TryParse(parts(0), copies) Then
                Dim copySet As List(Of String) = ReadTokens(parts(1))
                For i As Integer = 1 To copies
                    expandedToken.AddRange(copySet)
                Next
                didExpand = True
            ElseIf expandedToken IsNot Nothing AndAlso parts(0) = "!" Then
                expandedToken.AddRange(ReadTokens(parts(1)))
                remove = True
            End If
        End If

        Return token
    End Function

    Public Shared Function ReadToken(ByVal input As String, ByRef start As Integer, Optional ByRef expandedToken As List(Of String) = Nothing, Optional ByRef didExpand As Boolean = False, Optional ByRef remove As Boolean = False) As String
        If start = -1 Then
            Throw New ApplicationException("Read token past end of line")
        End If
        didExpand = False
        remove = False

        Dim len As Integer = input.Length

        If input(start) = """"c Then
            Dim origin As Integer = start + 1
            While start > -1 AndAlso start < len
                start = input.IndexOf(""""c, start + 1)

                If start = -1 Then
                    Exit While
                End If

                If input(start - 1) <> "\"c Then
                    Dim tokenlen As Integer = start - origin
                    start += 1

                    While start < len AndAlso input(start) = " "c
                        start += 1
                    End While

                    If start >= len Then
                        start = -1
                    ElseIf input.Substring(start).StartsWith("//") Then
                        start = -1
                    End If

                    Return input.Substring(origin, tokenlen).Replace("\""", """").Replace("\\", "\")
                End If
            End While

            Throw New ApplicationException("Read token past end of line. Missing a "", are we?")
        Else
            Dim origin As Integer = start

            start = input.IndexOf(" "c, start)

            If start > -1 Then
                Dim foundtoken As String = ExpandToken(input, origin, start, expandedToken, didExpand, remove)
                While start < len AndAlso input(start) = " "c
                    start += 1
                End While

                If start >= len OrElse input.Substring(start).StartsWith("//") Then
                    start = -1
                End If

                Return foundtoken
            ElseIf input.Last = "]"c Then
                Dim foundtoken As String = ExpandToken(input, origin, len, expandedToken, didExpand, remove)
                start = -1

                Return foundtoken
            Else
                Return input.Substring(origin)
            End If
        End If
    End Function

    Private Shared Sub ReadProject(ByVal project As Project, ByVal reader As IO.TextReader)
        If Not ReadLines(project, reader, AddressOf ParseProject) Then
            project.Warnings.Add("The code was not read to the end of the document")
        End If

        If project.StartRules.Count = 0 Then
            project.StartRules.Add("root", 100)
        End If

        For Each sr As String In project.StartRules.Keys
            If project.Rules.GetRuleByName(sr) Is Nothing Then
                project.Warnings.Add(String.Format("The starting rule '{0}' does not exist.", sr))
            End If
        Next

        For Each r As Rule In project.Rules
            For Each c As CommandBase In r.Commands
                c.CheckSanity(project)
            Next
        Next
    End Sub

    Private Shared Function ParseProject(ByVal context As Object, ByVal reader As IO.TextReader, ByVal line As String, ByRef lineNumber As Integer) As Boolean
        Dim project As Project = TryCast(context, Project)

        Dim start As Integer = 0

        Try
            If Not String.IsNullOrEmpty(line) Then
                Dim command = ReadToken(line, start)

                If command.Equals("tokens", StringComparison.CurrentCultureIgnoreCase) Then
                    Dim tks As New TokenSet()
                    Dim len As Integer = line.Length
                    Dim token As String = ReadToken(line, start)
                    tks.Name = token

                    Dim expand As New List(Of String)()
                    Dim didExpand As Boolean = False
                    Dim remove As Boolean = False
                    While start > -1 AndAlso start < len
                        token = ReadToken(line, start, expand, didExpand, remove)

                        If Not didExpand AndAlso Not remove Then
                            If token.StartsWith("$") AndAlso token.Length > 1 Then
                                Dim refTks As TokenSet = project.TokenSets.Where(Function(ts As TokenSet) ts.Name = token.Substring(1)).FirstOrDefault()
                                If refTks IsNot Nothing Then
                                    tks.Tokens.AddRange(refTks.Tokens)
                                Else
                                    project.Warnings.Add(String.Format("Line {0}: The Tokens directive referenced another token set '{1}' which did not exist at the time of parsing.", lineNumber, token.Substring(1)))
                                End If
                            Else
                                tks.Tokens.Add(token)
                            End If
                        End If
                    End While

                    If remove Then
                        tks.Tokens.RemoveAll(Function(tok As String) expand.Contains(tok))
                    Else
                        tks.Tokens.AddRange(expand)
                    End If

                    If tks.Tokens.Count = 0 Then
                        project.Warnings.Add(String.Format("Line {0}: The Tokens directive requires at least 2 arguments.", lineNumber))
                    End If

                    project.TokenSets.Add(tks)
                ElseIf command.Equals("rule", StringComparison.CurrentCultureIgnoreCase) Then
                    Dim r As New Rule()
                    r.Name = ReadToken(line, start)
                    r.LineNumber = lineNumber
                    Dim probability As Double = 1

                    Dim probstring As String = ReadToken(line, start)
                    If probstring = "{" Then
                        If start > -1 Then
                            Throw New ApplicationException("The Rule directive must end with a {")
                        End If
                    Else
                        If Not Double.TryParse(probstring, probability) Then
                            project.Warnings.Add(String.Format("Line {0}: The Rule directive requires the second argument to be numeric.", lineNumber))
                        End If

                        probstring = ReadToken(line, start)

                        If probstring <> "{" Then
                            Throw New ApplicationException("The Rule directive must end with a {")
                        End If
                    End If

                    r.Probability = probability

                    project.Rules.Add(r)
                    If ReadLines(New Object() {project, r.Commands}, reader, AddressOf ParseCommands, lineNumber) Then
                        project.Warnings.Add(String.Format("Line {1}: The rule '{0}' is not closed correctly.", r.Name, lineNumber))
                    End If
                ElseIf command.Equals("startingrule", StringComparison.CurrentCultureIgnoreCase) Then
                    Dim rulename As String = ReadToken(line, start)

                    If start > -1 Then
                        Dim amountstring As String = ReadToken(line, start)

                        If start > -1 Then
                            project.Warnings.Add(String.Format("Line {0}: The StartingRule directive requires 1 or 2 arguments.", lineNumber))
                        End If

                        Dim val As Integer
                        If Integer.TryParse(amountstring, val) Then
                            project.StartRules(rulename) = val
                        Else
                            project.Warnings.Add(String.Format("Line {0}: The StartingRule directive requires that the second argument is a number.", lineNumber))
                        End If
                    Else
                        project.StartRules(rulename) = 100
                    End If
                ElseIf command.Equals("Column", StringComparison.CurrentCultureIgnoreCase) Then
                    Dim name As String = ReadToken(line, start)
                    If start > -1 Then
                        Dim value As String = ReadToken(line, start)

                        project.Columns.Add(name, value)
                    Else
                        project.Warnings.Add(String.Format("Line {0}: The Column directive requires two arguments.", lineNumber))
                    End If
                Else
                    project.Warnings.Add(String.Format("Line {1}: The command '{0}' was not recognized.", command, lineNumber))
                End If
            End If
        Catch ex As Exception
            project.Warnings.Add(String.Format("Line {0}: {1}", lineNumber, ex.Message))
        End Try

        Return True
    End Function

    Private Shared s_Commands As List(Of Type)
    Private Shared ReadOnly Property Commands() As List(Of Type)
        Get
            If s_Commands Is Nothing Then
                s_Commands = (From t As Type In Reflection.Assembly.GetAssembly(GetType(ProjectSerializer)).GetTypes() Where t.IsSubclassOf(GetType(CommandBase))).ToList()
            End If
            Return s_Commands
        End Get
    End Property

    Public Shared Function ParseWeightedCommands(ByVal context As Object, ByVal reader As IO.TextReader, ByVal line As String, ByRef lineNumber As Integer) As Boolean
        If String.IsNullOrEmpty(line) Then
            Return True
        End If

        If line.StartsWith("}") Then
            Return False
        End If

        Dim args() As Object = TryCast(context, Object())
        Dim project As Project = TryCast(args(0), Project)
        Dim cmds As List(Of WeightedCommand) = TryCast(args(1), List(Of WeightedCommand))

        Dim start As Integer = 0
        Dim weightStr As String = ReadToken(line, start)

        Dim cmdStart As Integer = start

        Dim weight As Double = 0D
        If Not Double.TryParse(weightStr, weight) Then
            project.Warnings.Add(String.Format("Line {0}: Expected a weighted command.", lineNumber))
        Else
            Dim command As String = ReadToken(line, start)

            Dim commandCandidates As IEnumerable(Of Type) = From cmd As Type In Commands Where cmd.Name.ToLower().StartsWith(command.ToLower())

            If commandCandidates.Count > 1 Then
                project.Warnings.Add(String.Format("Line {1}: The command '{0}' is ambiguous.", command, lineNumber))
            Else
                Dim tp As Type = commandCandidates.FirstOrDefault()

                If tp IsNot Nothing Then
                    Dim c As CommandBase = TryCast(Activator.CreateInstance(tp), CommandBase)

                    If c IsNot Nothing Then
                        c.LoadCommand(project, reader, line.Substring(cmdStart), lineNumber)
                    Else
                        project.Warnings.Add(String.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber))
                    End If

                    cmds.Add(New WeightedCommand() With {.Command = c, .Weight = weight})
                Else
                    project.Warnings.Add(String.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber))
                End If
            End If
        End If

        Return True
    End Function

    Public Shared Function ParseCommands(ByVal context As Object, ByVal reader As IO.TextReader, ByVal line As String, ByRef lineNumber As Integer) As Boolean
        If String.IsNullOrEmpty(line) Then
            Return True
        End If

        If line.StartsWith("}") Then
            Return False
        End If

        Dim args() As Object = TryCast(context, Object())
        Dim project As Project = TryCast(args(0), Project)
        Dim cmds As CommandCollection = TryCast(args(1), CommandCollection)

        Dim start As Integer = 0
        Dim command As String = ReadToken(line, start)

        Dim commandCandidates As IEnumerable(Of Type) = From cmd As Type In Commands Where cmd.Name.ToLower().StartsWith(command.ToLower())

        If commandCandidates.Count > 1 Then
            project.Warnings.Add(String.Format("Line {1}: The command '{0}' is ambiguous.", command, lineNumber))
        Else
            Dim tp As Type = commandCandidates.FirstOrDefault()

            If tp IsNot Nothing Then
                Dim c As CommandBase = TryCast(Activator.CreateInstance(tp), CommandBase)

                If c IsNot Nothing Then
                    c.LoadCommand(project, reader, line, lineNumber)
                Else
                    project.Warnings.Add(String.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber))
                End If

                cmds.Add(c)
            Else
                project.Warnings.Add(String.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber))
            End If
        End If

        Return True
    End Function

    Public Shared Sub SaveProject(ByVal project As Project, ByVal path As String)
        Using tw As New IO.StreamWriter(path, False, Text.Encoding.UTF8)
            WriteProject(project, tw)
        End Using
    End Sub

    Private Shared Sub WriteProject(ByVal project As Project, ByVal writer As IO.TextWriter)
        For Each startRule As String In project.StartRules.Keys
            writer.WriteLine(String.Format("StartingRule {0} {1}", startRule, project.StartRules(startRule)))
        Next

        For Each ts As TokenSet In project.TokenSets
            writer.WriteLine(String.Format("Tokens {0} {1}", ts.Name, String.Join(" ", ts.Tokens.ToArray())))
        Next

        For Each r As Rule In project.Rules
            writer.WriteLine(String.Format("Rule {0} {1} {{}", r.Name, r.Probability))
            r.Commands.WriteCommands(writer)
            writer.WriteLine("}")
        Next
    End Sub

    Public Shared Function SecureString(ByVal input As String) As String
        If input.Contains(" ") Then
            Return String.Format("""{0}""", input)
        End If

        Return input
    End Function

    Public Shared Function SecureList(ByVal list As IEnumerable(Of String)) As String
        Dim sb As New System.Text.StringBuilder()

        For Each token As String In list
            sb.Append(" ")
            sb.Append(SecureString(token))
        Next

        Return sb.ToString().TrimStart()
    End Function
End Class

File: monowordbuilder/wordbuilderbase/Project/ProjectSerializer.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Reflection;
using Whee.WordBuilder.Helpers;
using Whee.WordBuilder.Model.Commands;

public delegate bool LineParserDelegate(object context, TextReader reader, string line, ref int lineNumber);

namespace Whee.WordBuilder.Model
{
	public sealed class ProjectSerializer
	{
		private ProjectSerializer()
		{
		}
			
		public static bool ReadLines(object context, TextReader reader, LineParserDelegate parser, ref int lineNumber)
		{
			string line = reader.ReadLine();
			while (line != null) {
				line = line.Trim(' ', '\x09');
				lineNumber += 1;
				if (line.StartsWith("/*")) {
					int startingLine = lineNumber;
					while (line != null && !line.EndsWith("*/")) {
						line = reader.ReadLine();
						lineNumber += 1;
					}
					if (line == null) {
						throw new ApplicationException(string.Format("Line {0}: Block comment not closed correctly.", startingLine));
					}
				}
				else if (line.StartsWith("//")) {
				}
				// Ignore single line comments.
				else if (!parser.Invoke(context, reader, line.Trim(' ', '\x09'), ref lineNumber)) {
					return false;
				}
				line = reader.ReadLine();
			}
	
			return true;
		}
	
		public static List<string> ReadTokens(string input)
		{
			List<string> result = new List<string>();
			int start = 0;
			List<string> expand = new List<string>();
			bool didExpand = false;
			bool doRemove = false;
			
			while (start > -1) {
				string token = ReadToken(input, ref start, ref expand, ref didExpand, ref doRemove);
	
				if (didExpand) {
					result.AddRange(expand);
					expand.Clear();
				}
				else {
					result.Add(token);
				}
			}
	
			return result;
		}
	
		public static string ExpandToken(string input, int origin, ref int start, ref List<string> expandedToken, ref bool didExpand, ref bool @remove)
		{
			string token = input.Substring(origin, start - origin);
	
			if (token.Contains("[")) {
				int nextOpen = input.IndexOf('[', start);
				start = input.IndexOf(']', start - 1) + 1;
	
				while (nextOpen > -1 && nextOpen < start - 1) {
					nextOpen = input.IndexOf('[', nextOpen + 1);
					start = input.IndexOf(']', start) + 1;
				}
	
				if (start == 0) {
					throw new ApplicationException("Read token past end of line. Missing a ], are we?");
				}
	
				token = input.Substring(origin, start - origin);
				string[] parts = token.Substring(0, token.Length - 1).Split(new char[] { '[' }, 2);
	
				int copies = 0;
				if (expandedToken != null && int.TryParse(parts[0], out copies)) {
					List<string> copySet = ReadTokens(parts[1]);
					for (int i = 1; i <= copies; i++) {
						expandedToken.AddRange(copySet);
					}
					didExpand = true;
				}
				else if (expandedToken != null && parts[0] == "!") {
					expandedToken.AddRange(ReadTokens(parts[1]));
					@remove = true;
				}
			}
	
			return token;
		}
	
		public static string ReadToken(string input, ref int start)
		{
			List<string> expandedToken = null;
			bool didExpand = false;
			bool doRemove = false;
			return ReadToken(input, ref start, ref expandedToken, ref didExpand, ref doRemove);
		}
		
		public static string ReadToken(string input, ref int start,ref List<string> expandedToken, ref bool didExpand, ref bool @remove)
		{
			if (start == -1) {
				throw new ApplicationException("Read token past end of line");
			}
			didExpand = false;
			@remove = false;
	
			int len = input.Length;
	
			if (input[start] == '"') {
				int origin = start + 1;
				while (start > -1 && start < len) {
					start = input.IndexOf('"', start + 1);
	
					if (start == -1) {
						break; // TODO: might not be correct. Was : Exit While
					}
	
					if (input[start - 1] != '\\') {
						int tokenlen = start - origin;
						start += 1;
	
						while (start < len && input[start] == ' ') {
							start += 1;
						}
	
						if (start >= len) {
							start = -1;
						}
						else if (input.Substring(start).StartsWith("//")) {
							start = -1;
						}
	
						return input.Substring(origin, tokenlen).Replace("\\\"", "\"").Replace("\\\\", "\\");
					}
				}
	
				throw new ApplicationException("Read token past end of line. Missing a \", are we?");
			}
			else {
				int origin = start;
	
				start = input.IndexOf(' ', start);
	
				if (start > -1) {
					string foundtoken = ExpandToken(input, origin, ref start, ref expandedToken, ref didExpand, ref @remove);
					while (start < len && input[start] == ' ') {
						start += 1;
					}
	
					if (start >= len || input.Substring(start).StartsWith("//")) {
						start = -1;
					}
	
					return foundtoken;
				}
				else if (input[input.Length - 1] == ']') {
					string foundtoken = ExpandToken(input, origin, ref len, ref expandedToken, ref didExpand, ref @remove);
					start = -1;
	
					return foundtoken;
				}
				else {
					return input.Substring(origin);
				}
			}
		}
	
		private static void ReadProject(Project project, TextReader reader)
		{
            //int linenumber = 0;
			
            //if (!ReadLines(project, reader, ParseProject, ref linenumber)) {
            //    project.Warnings.Add("The code was not read to the end of the document");
            //}
	
            //if (project.StartRules.Count == 0) {
            //    project.StartRules.Add("root", 100);
            //}
	
            //foreach (string sr in project.StartRules.Keys) {
            //    if (project.Rules.GetRuleByName(sr) == null) {
            //        project.Warnings.Add(string.Format("The starting rule '{0}' does not exist.", sr));
            //    }
            //}
	
            //foreach (Rule r in project.Rules) {
            //    foreach (CommandBase c in r.Commands) {
            //        c.CheckSanity(project);
            //    }
            //}
		}
	
		private static bool ParseProject(object context, TextReader reader, string line, ref int lineNumber)
		{
			Project project = context as Project;
	
			int start = 0;
	
			try {
				if (!string.IsNullOrEmpty(line)) {
					string command = ReadToken(line, ref start);
	
					if (command.Equals("tokens", StringComparison.CurrentCultureIgnoreCase)) {
                        TokenSet tks = new TokenSet(null, Whee.WordBuilder.Helpers.Random.Instance);
						int len = line.Length;
						string token = ReadToken(line, ref start);
						tks.Name = token;
	
						List<string> expand = new List<string>();
						bool didExpand = false;
						bool @remove = false;
						while (start > -1 && start < len) {
							token = ReadToken(line, ref start, ref expand, ref didExpand, ref @remove);
	
							if (!didExpand && !@remove) {
								if (token.StartsWith("$") && token.Length > 1) {
									TokenSet refTks = project.TokenSets.FindByName(token.Substring(1));
									if (refTks != null) {
										tks.Tokens.AddRange(refTks.Tokens);
									}
									else {
										project.Warnings.Add(string.Format("Line {0}: The Tokens directive referenced another token set '{1}' which did not exist at the time of parsing.", lineNumber, token.Substring(1)));
									}
								}
								else {
									tks.Tokens.Add(token);
								}
							}
						}
	
						if (@remove) {
							tks.Tokens.RemoveAll((string tok) => expand.Contains(tok));
						}
						else {
							tks.Tokens.AddRange(expand);
						}
	
						if (tks.Tokens.Count == 0) {
							project.Warnings.Add(string.Format("Line {0}: The Tokens directive requires at least 2 arguments.", lineNumber));
						}
	
						project.TokenSets.Add(tks);
					}
					else if (command.Equals("rule", StringComparison.CurrentCultureIgnoreCase)) {
						Rule r = new Rule(null);
						r.Name = ReadToken(line, ref start);
						r.LineNumber = lineNumber;
						double probability = 1;
	
						string probstring = ReadToken(line, ref start);
						if (probstring == "{") {
							if (start > -1) {
								throw new ApplicationException("The Rule directive must end with a {");
							}
						}
						else {
							if (!double.TryParse(probstring, out probability)) {
								project.Warnings.Add(string.Format("Line {0}: The Rule directive requires the second argument to be numeric.", lineNumber));
							}
	
							probstring = ReadToken(line, ref start);
	
							if (probstring != "{") {
								throw new ApplicationException("The Rule directive must end with a {");
							}
						}
	
						r.Probability = probability;
	
						project.Rules.Add(r);
						if (ReadLines(new object[] { project, r.Commands }, reader, ParseCommands, ref lineNumber)) {
							project.Warnings.Add(string.Format("Line {1}: The rule '{0}' is not closed correctly.", r.Name, lineNumber));
						}
					}
					else if (command.Equals("startingrule", StringComparison.CurrentCultureIgnoreCase)) {
						string rulename = ReadToken(line, ref start);
	
						if (start > -1) {
							string amountstring = ReadToken(line, ref start);
	
							if (start > -1) {
								project.Warnings.Add(string.Format("Line {0}: The StartingRule directive requires 1 or 2 arguments.", lineNumber));
							}
	
							int val = 0;
							if (int.TryParse(amountstring, out val)) {
								project.StartRules[rulename] = val;
							}
							else {
								project.Warnings.Add(string.Format("Line {0}: The StartingRule directive requires that the second argument is a number.", lineNumber));
							}
						}
						else {
							project.StartRules[rulename] = 100;
						}
					}
					else if (command.Equals("Column", StringComparison.CurrentCultureIgnoreCase)) {
						string name = ReadToken(line, ref start);
						if (start > -1) {
							string value = ReadToken(line, ref start);
	
							project.Columns.Add(name, value);
						}
						else {
							project.Warnings.Add(string.Format("Line {0}: The Column directive requires two arguments.", lineNumber));
						}
					}
					else {
						project.Warnings.Add(string.Format("Line {1}: The command '{0}' was not recognized.", command, lineNumber));
					}
				}
			}
			catch (Exception ex) {
				project.Warnings.Add(string.Format("Line {0}: {1}", lineNumber, ex.Message));
			}
	
			return true;
		}
	
		private static List<Type> s_Commands;
		private static List<Type> Commands {
			get {
				if (s_Commands == null) {
					s_Commands = new List<Type>();
					foreach (Type t in Assembly.GetAssembly(typeof(ProjectSerializer)).GetTypes()) {
						if (t.IsSubclassOf(typeof(CommandBase))) {
							s_Commands.Add(t);
						}
					}
				}
				return s_Commands;
			}
		}
	
		public static bool ParseWeightedCommands(object context, TextReader reader, string line, ref int lineNumber)
		{
			if (string.IsNullOrEmpty(line)) {
				return true;
			}
	
			if (line.StartsWith("}")) {
				return false;
			}
	
			object[] args = context as object[];
			Project project = args[0] as Project;
			List<WeightedCommand> cmds = args[1] as List<WeightedCommand>;
	
			int start = 0;
			string weightStr = ReadToken(line, ref start);
	
			int cmdStart = start;
	
			double weight = 0;
			if (!double.TryParse(weightStr, out weight)) {
				project.Warnings.Add(string.Format("Line {0}: Expected a weighted command.", lineNumber));
			}
			else {
				string command = ReadToken(line, ref start);
	
				List<Type> commandCandidates = new List<Type>();
				foreach (Type cmd in Commands) {
					if (cmd.Name.ToLower().StartsWith(command.ToLower())) {
						commandCandidates.Add(cmd);
					}
				}
	
				if (commandCandidates.Count > 1) {
					project.Warnings.Add(string.Format("Line {1}: The command '{0}' is ambiguous.", command, lineNumber));
				}
				else {
					Type tp = commandCandidates.Count > 0 ? commandCandidates[0] : null;
	
					if (tp != null) {
						CommandBase c = Activator.CreateInstance(tp) as CommandBase;
	
						if (c != null) {
							c.LoadCommand(project, reader, line.Substring(cmdStart), ref lineNumber);
						}
						else {
							project.Warnings.Add(string.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber));
						}
	
						WeightedCommand wc = new WeightedCommand();
						wc.Command = c;
						wc.Weight = weight;
						cmds.Add(wc);
					}
					else {
						project.Warnings.Add(string.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber));
					}
				}
			}
	
			return true;
		}
	
		public static bool ParseCommands(object context, TextReader reader, string line, ref int lineNumber)
		{
			if (string.IsNullOrEmpty(line)) {
				return true;
			}
	
			if (line.StartsWith("}")) {
				return false;
			}
	
			object[] args = context as object[];
			Project project = args[0] as Project;
			CommandCollection cmds = args[1] as CommandCollection;
	
			int start = 0;
			string command = ReadToken(line, ref start);
	
			List<Type> commandCandidates = new List<Type>();
			foreach (Type cmd in Commands) {
				if (cmd.Name.ToLower().StartsWith(command.ToLower())) {
					commandCandidates.Add(cmd);
				}
			}
	
			if (commandCandidates.Count > 1) {
				project.Warnings.Add(string.Format("Line {1}: The command '{0}' is ambiguous.", command, lineNumber));
			}
			else {
				Type tp = commandCandidates.Count > 0 ? commandCandidates[0] : null;
	
				if (tp != null) {
					CommandBase c = Activator.CreateInstance(tp) as CommandBase;
	
					if (c != null) {
						c.LoadCommand(project, reader, line, ref lineNumber);
					}
					else {
						project.Warnings.Add(string.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber));
					}
	
					cmds.Add(c);
				}
				else {
					project.Warnings.Add(string.Format("Line {1}: The command '{0}' is not valid.", command, lineNumber));
				}
			}
	
			return true;
		}
	
		public static void SaveProject(Project project, string path)
		{
			using (StreamWriter tw = new StreamWriter(path, false, System.Text.Encoding.UTF8)) {
				WriteProject(project, tw);
			}
		}
	
		private static void WriteProject(Project project, TextWriter writer)
		{
			foreach (string startRule in project.StartRules.Keys) {
				writer.WriteLine(string.Format("StartingRule {0} {1}", startRule, project.StartRules[startRule]));
			}
	
			foreach (TokenSet ts in project.TokenSets) {
				writer.WriteLine(string.Format("Tokens {0} {1}", ts.Name, string.Join(" ", ts.Tokens.ToArray())));
			}
	
			foreach (Rule r in project.Rules) {
				writer.WriteLine(string.Format("Rule {0} {1} {{}", r.Name, r.Probability));
				r.Commands.WriteCommands(writer);
				writer.WriteLine("}");
			}
		}
	
		public static string SecureString(string input)
		{
			if (input.Contains(" ")) {
				return string.Format("\"{0}\"", input);
			}
	
			return input;
		}
	
		public static string SecureList(IEnumerable<string> list)
		{
			System.Text.StringBuilder sb = new System.Text.StringBuilder();
	
			foreach (string token in list) {
				sb.Append(" ");
				sb.Append(SecureString(token));
			}
	
			return sb.ToString().TrimStart();
		}
	}
}